packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts

```diff
@@ -158,48 +158,14 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
             },
           ],
         };
-      } else if (
-        // source exists
-        Boolean(indices[source!]?.mappings?._meta?.migrationMappingPropertyHashes) &&
-        // ...and mappings are unchanged
-        !diffMappings(
-          /* actual */
-          indices[source!].mappings,
-          /* expected */
-          stateP.targetIndexMappings
-        )
-      ) {
-        const targetIndex = source!;
-        const sourceMappings = indices[source!].mappings;
-
-        return {
-          ...stateP,
-          controlState: 'PREPARE_COMPATIBLE_MIGRATION',
-          sourceIndex: Option.none,
-          targetIndex,
-          targetIndexRawMappings: sourceMappings,
-          targetIndexMappings: mergeMigrationMappingPropertyHashes(
-            stateP.targetIndexMappings,
-            sourceMappings
-          ),
-          preTransformDocsActions: [
-            // Point the version alias to the source index. This let's other Kibana
-            // instances know that a migration for the current version is "done"
-            // even though we may be waiting for document transformations to finish.
-            { add: { index: source!, alias: stateP.versionAlias } },
-            ...buildRemoveAliasActions(source!, Object.keys(aliases), [
-              stateP.currentAlias,
-              stateP.versionAlias,
-            ]),
-          ],
-          versionIndexReadyActions: Option.none,
-        };
       } else if (
         // If the `.kibana` alias exists
         source != null
       ) {
+        // CHECKPOINT here we decide to go for yellow source
         return {
           ...stateP,
+          aliases,
```

Comment:
Fair point!

The only way I see to get fresher _aliases_ information would be to perform a `GET .kibana` again (like the `INIT` step does), and even in that case, there's no way to guarantee that another instance will not delete the aliases before us.

- If the other instance is the same version, and it performs the same cleanup of aliases before us, it's alright to proceed to `OUTDATED_DOCUMENTS_SEARCH_OPEN_PIT` on `alias_not_found`. As you said, odds of this happening are increasing with this fix, but the still behavior / handling is correct IMHO.

- If the other instance is a more recent version, it will obtain the aliases in the INIT step:
  - if it obtains them **before** we update / delete them in the PREPARE_COMPATIBLE_MIGRATION step, it will proceed to CHECK_UNKNOWN_DOCUMENTS + all the reindex steps, which don't use the aliases anymore.
  - if it obtains them **after** we update / delete them in the PREPARE_COMPATIBLE_MIGRATION step, it will reindex the source index into a new index without using the alias for anything. Then, it would likely override our `.kibana` alias and make it point to the newly created index. In this case we'd have some garbage version aliases pointing to an older index version, but that's already the case without this fix.

Reply:
++ there's always a potential race condition between the time we get the fetchAliases respose and performing the `preTransformDocsActions`. Before it was probably in the order of < 100ms and now it could potentially be many minutes. So this makes any potential race conditions more likely to occur but doesn't introduce anything new.

Like you both said, any race conditions should be handled in `PREPARE_COMPATIBLE_MIGRATION` already.